arangod/RestHandler/RestDumpHandler.cpp

```diff
@@ -0,0 +1,269 @@
+  if (base.ok() && !std::get<0>(base.get()).empty()) {
+    return base;
+  }
+
+  auto type = _request->requestType();
+  auto const& suffixes = _request->suffixes();
+  size_t const len = suffixes.size();
+  if (type == rest::RequestType::DELETE_REQ) {
+    if (len != 1) {
+      return ResultT<std::pair<std::string, bool>>::error(
+          TRI_ERROR_BAD_PARAMETER, "expecting DELETE /_api/dump/<id>");
+    }
+  } else if (type == rest::RequestType::POST) {
+    if ((len == 1 && suffixes[0] != "start") ||
+        (len == 2 && suffixes[0] != "next") || len < 1 || len > 2) {
+      return ResultT<std::pair<std::string, bool>>::error(
+          TRI_ERROR_BAD_PARAMETER,
+          "expecting POST /_api/dump/start or /_api/dump/next/<id>");
+    }
+  } else {
+    // invalid HTTP method
+    return ResultT<std::pair<std::string, bool>>::error(
+        TRI_ERROR_HTTP_METHOD_NOT_ALLOWED);
+  }
+
+  if (ServerState::instance()->isCoordinator()) {
+    ServerID const& DBserver = _request->value("dbserver");
+    if (!DBserver.empty()) {
+      // if DBserver property present, remove auth header
+      return std::make_pair(DBserver, true);
+    }
+
+    return ResultT<std::pair<std::string, bool>>::error(
+        TRI_ERROR_BAD_PARAMETER, "need a 'dbserver' parameter");
+  }
+
+  return {std::make_pair(StaticStrings::Empty, false)};
+}
+
+void RestDumpHandler::handleCommandDumpStart() {
+  std::string database = _request->databaseName();
+  std::string user = _request->user();
+
+  // TODO: check permissions
+
+  bool parseSuccess = false;
+  VPackSlice body = this->parseVPackBody(parseSuccess);
+  if (!parseSuccess) {  // error message generated in parseVPackBody
+    return;
+  }
+
+  if (!body.isObject()) {
+    generateError(Result(TRI_ERROR_BAD_PARAMETER, "invalid request body"));
+    return;
+  }
+
+  RocksDBDumpContextOptions opts;
+  opts.batchSize = [&body]() {
+    if (auto s = body.get("batchSize"); s.isNumber()) {
+      return s.getNumber<uint64_t>();
+    }
+    return ::defaultBatchSize;
+  }();
+  opts.prefetchCount = [&body]() {
+    if (auto s = body.get("prefetchCount"); s.isNumber()) {
+      return s.getNumber<uint64_t>();
+    }
+    return ::defaultPrefetchCount;
+  }();
+  opts.parallelism = [&body]() {
+    if (auto s = body.get("parallelism"); s.isNumber()) {
+      return s.getNumber<uint64_t>();
+    }
+    return ::defaultParallelism;
+  }();
+  opts.ttl = [&body]() {
+    if (auto s = body.get("ttl"); s.isNumber()) {
+      return s.getNumber<double>();
+    }
+    return ::defaultTtl;
+  }();
+
+  if (auto s = body.get("shards"); !s.isArray()) {
+    generateError(
+        Result(TRI_ERROR_BAD_PARAMETER, "invalid 'shards' value in request"));
+    return;
+  } else {
+    for (auto it : VPackArrayIterator(s)) {
+      opts.shards.emplace_back(it.copyString());
+    }
+  }
+
+  auto& engine =
+      server().getFeature<EngineSelectorFeature>().engine<RocksDBEngine>();
+  auto* manager = engine.dumpManager();
```

Comment:
Not a fan of this whole `getFeature` stuff. Can we at least move this to the constructor and store the manager as a member, pretending that all dependencies have been passed in (constructor injection) instead of fetching them dynamically?

Reply:
I tried your suggestion in commit 0cd69ed8770d6845d35b10e722944cc9a6bc00fa, but it turned out that it didn't work.
It seems to be problematic to access the storage engine's `dumpManager()` function in the RestHandler's constructor, simply because the manager seemed to be a nullptr at that point.
This is probably related to some hot backup starting/restarting servers at will, but I don't know for sure.
The problem of the manager being a nullptr can be avoided by using the existing code, which only calls `dumpManager()` during execute().

Reply:
Ok, this is probably related to our lack of dependency management and the resulting initialization order. I would still prefer if we could at least move the getFeature call into the constructor and keep the reference to the engine as member and only defer the `dumpManager` call.